source/Calamari.Common/Features/StructuredVariables/XmlFormatVariableReplacer.cs

```diff
@@ -0,0 +1,234 @@
+            
+            var doc = new XmlDocument();
+            try
+            {
+                doc.LoadXml(fileText);
+            }
+            catch (XmlException e)
+            {
+                throw new StructuredConfigFileParseException(e.Message, e);
+            }
+
+            var nsManager = BuildNsManagerFromDocument(doc);
+            var navigator = doc.CreateNavigator();
+            
+            foreach (var variable in variables)
+            {
+                if (IsValidXPath(variable.Key, nsManager))
+                {
+                    var xPathExpression = variable.Key;
+                    var selectedNodes = navigator.XPath2SelectNodes(xPathExpression, nsManager);
+
+                    foreach (XPathNavigator selectedNode in selectedNodes)
+                    {
+                        switch (selectedNode.UnderlyingObject)
+                        {
+                            case XmlText text:
+                                text.Data = variable.Value;
+                                break;
+
+                            case XmlAttribute attribute:
+                                attribute.Value = variable.Value;
+                                break;
+
+                            case XmlComment comment:
+                                comment.Data = variable.Value;
+                                break;
+
+                            case XmlElement element:
+                                if (element.ChildNodes.Count == 1 && element.ChildNodes[0].NodeType == XmlNodeType.CDATA)
+                                {
+                                    // Try to preserve CDatas in the output.
+                                    element.ChildNodes[0].Value = variable.Value;
+                                }
+                                else if (ContainsElements(element))
+                                {
+                                    TrySetInnerXml(element, xPathExpression, variable.Value);
+                                }
+                                else
+                                {
+                                    element.InnerText = variable.Value;
+                                }
+                                break;
+
+                            case XmlCharacterData cData:
+                                cData.Data = variable.Value;
+                                break;
+
+                            case XmlProcessingInstruction processingInstruction:
+                                processingInstruction.Data = variable.Value;
+                                break;
+                            
+                            case XmlNode node:
+                                log.Warn($"Node of type {node.NodeType} not supported");
+                                break;
+
+                            default:
+                                // TODO: consider whether to silently ignore
```

Comment:
Agreed. I forgot to come back to this one. I've put in place some better handling for this scenario.

There's a bit of context for how this all works together:

1. XPaths in XPath 2.0 always return a sequence of things, but those things aren't always nodes (because XPath 2.0 supports things like calculations), and we can't perform replacement on something that's not a node.
2. On [this line](https://github.com/OctopusDeploy/Calamari/pull/628/files/0c45e986b9f83483b37119c48322ec09bb7bb437#diff-061eda38d8f1b947e74af6a65cb87c83R57), we specifically filter down the results to just nodes, because we're using `XPath2SelectNodes` and not `XPath2Select`.
3. So if I'm understanding all of that correctly, then the default case shouldn't get hit, but I like having handling for it in case of bugs.

I've also changed the handling for this case to warn rather than fail. Probably doesn't matter either way though to be fair.